### ISSUE TITLE
Using "·" as non-breaking space fails on 1.1.0 (works on 1.0.0 and 1.0.1)

### ISSUE THREAD

**Issue by jrgonzalezg** [square/kotlinpoet#643]
I am trying to migrate Arrow from KotlinPoet 1.0.0-RC1 to 1.1.0 (https://github.com/arrow-kt/arrow/pull/1345) and I got into an issue with new line wraping on spaces so I identified the places where that happened and replaced the spaces with `·` (https://github.com/arrow-kt/arrow/pull/1345/commits/130353cca59c3237a9bc7472ac37a48f893d31ee) which is supposed to produce a non-breaking space.

While this works if I only update KotlinPoet version to 1.0.0 or 1.0.1 it fails again due to line wrapping when using 1.1.0 so it seems like some kind of regression. The version used on the Arrow PR can be changed on the main build.gradle file to reproduce the issue if you want. It is set to 1.1.0 (that fails) but it succeeds if rolled back to 1.0.0 or 1.0.1 (no code changes needed).

As an example here is some of the code generated for `modules/core/arrow-core-extensions/build/generated/source/kaptKotlin/main/extension/arrow/core/extensions/const/functor/C
onstFunctor.kt` with KotlinPoet 1.0.1:
```
fun <A> Kind<Kind<ForConst, A>, A>.unit(): Const<A, Unit> =
        arrow.typeclasses.Const.functor<A>().run {
  this@unit.unit() as arrow.typeclasses.Const<A, kotlin.Unit>
}
```

When switching to 1.1.0 the same code becomes:
```
fun <A> Kind<Kind<ForConst, A>, A>.unit(): Const<A, Unit> = arrow.typeclasses.Const.functor<A>().run
        {
  this@unit.unit() as arrow.typeclasses.Const<A, kotlin.Unit>
}
```

which doesn't compile due to the new line before the brace.

**Comment by Egorand** [comment]
A reproducible test case would really help figure out what's broken.

Anyway, here's my attempt to generate the example code you shared:
```kotlin
@Test fun runBraces() {
  val a = TypeVariableName("A")
  val kind = ClassName("arrow.typeclasses", "Kind")
  val forConst = ClassName("arrow.typeclasses", "ForConst")
  val const = ClassName("arrow.typeclasses", "Const")
  val constAUnit = const.parameterizedBy(a, UNIT)
  val functor = MemberName(const, "functor<A>")
  val file = FileSpec.builder("test", "Test")
      .addFunction(FunSpec.builder("unit")
          .addTypeVariable(a)
          .receiver(kind.parameterizedBy(kind.parameterizedBy(forConst, a), a))
          .returns(constAUnit)
          .beginControlFlow("return %M().run", functor)
          .addStatement("this@unit.unit() as %T", constAUnit)
          .endControlFlow()
          .build())
      .build()
  assertThat(file.toString()).isEqualTo("""
    |package test
    |
    |import arrow.typeclasses.Const
    |import arrow.typeclasses.Const.functor<A>
    |import arrow.typeclasses.ForConst
    |import arrow.typeclasses.Kind
    |import kotlin.Unit
    |
    |fun <A> Kind<Kind<ForConst, A>, A>.unit(): Const<A, Unit> = functor<A>().run {
    |    this@unit.unit() as Const<A, Unit>
    |}
    |""".trimMargin())
}
```

**Comment by jrgonzalezg** [comment]
Ok, sorry it took me quite a while to figure out how to make a minimal example since I was not familiar with annotation processors and that part of Arrow's code and I needed to learn how to connect to the annotation processor with a debugger and look at the exact parameters passed. I then minimized the test code to just this which already shows the issue:
```
  @Test
  fun runBracesFailing() {
    val file = FileSpec.builder("test", "Test")
        .addFunction(
            FunSpec.builder("unit")
                .addCode(CodeBlock.of(
                    "return arrow.typeclasses.Const.long.line.long.line.long.line.long.line.long.line.long.line.functor<A>().run {\n  this@unit.unit() as arrow.typeclasses.Const<A, kotlin.Unit>\n}\n"
                )
                )
                .build()
        )
        .build()

    assertThat(file.toString()).isEqualTo(
        """
    |package test
    |
    |fun unit() =
    |        arrow.typeclasses.Const.long.line.long.line.long.line.long.line.long.line.long.line.functor<A>().run {
    |  this@unit.unit() as arrow.typeclasses.Const<A, kotlin.Unit>
    |}
    |""".trimMargin()
    )
  }
```

Hope this helps to track the issue. I also noticed that removing the FileSpec the function code generated is quite different (no line wrapping):
```
fun unit() = arrow.typeclasses.Const.long.line.long.line.long.line.long.line.long.line.long.line.functor<A>().run {
  this@unit.unit() as arrow.typeclasses.Const<A, kotlin.Unit>
}
```

so triggering the issue seems to require some interaction with the `FileSpec` and the `FunSpec`.

**Comment by jrgonzalezg** [comment]
Oops, sorry, the minimal example is wrong, forgot the use of `·` when tweaking it 😅. I will try to get something else.

**Comment by jrgonzalezg** [comment]
I have continued trying things and now I think it is not strictly any issue with Kotlinpoet, but the usage Arrow was making of it. In particular I think this is what breaks Arrow: https://github.com/square/kotlinpoet/pull/529

So the problem is not that "Using `·` as non-breaking space does not work" but that previously you could go back and forth from a CodeBlock and a String and that `·` was preserved, but the new version of `toString()` removes the `·` and that breaks Arrow cause while it initially starts with a non-breaking space when it gets actual code generation now the `·` is no longer there but instead we have a normal space now.

Considering that there is probably nothing to change / fix on Kotlinpoet itself since the issue is probably that Arrow should not have relied on the previous `toString()` behavior, so I'm closing this issue. Thanks for the tips on how to make the minimal example that helped me on tracking it down :)

**Comment by Egorand** [comment]
You're welcome, thanks for digging into this!

It's true that the behavior of `toString()` on `*Spec` classes is different from what you get when pushing them into a `FileSpec`. `toString()` sets line length limit to `Int.MAX_VALUE`, hence you'll never face line wrapping issues.

On a side note, there's been a recent change (#641, available in 1.2.0-SNAPSHOT) that ensures the opening brace doesn't get wrapped when you use `beginControlFlow()` or `nextControlFlow()` on either a `CodeBlock` or a `FunSpec` - this is what I'd recommend using for cases such as above.
